api/grpcserver/v2alpha1/reward.go

```diff
@@ -0,0 +1,343 @@
+	RewardStream = "reward_stream_v2alpha1"
+)
+
+func NewRewardStreamService(db sql.Executor) *RewardStreamService {
+	return &RewardStreamService{db: db}
+}
+
+type RewardStreamService struct {
+	db sql.Executor
+}
+
+func (s *RewardStreamService) RegisterService(server *grpc.Server) {
+	spacemeshv2alpha1.RegisterRewardStreamServiceServer(server, s)
+}
+
+func (s *RewardStreamService) RegisterHandlerService(mux *runtime.ServeMux) error {
+	return spacemeshv2alpha1.RegisterRewardStreamServiceHandlerServer(context.Background(), mux, s)
+}
+
+func (s *RewardStreamService) Stream(
+	request *spacemeshv2alpha1.RewardStreamRequest,
+	stream spacemeshv2alpha1.RewardStreamService_StreamServer,
+) error {
+	ctx := stream.Context()
+	var sub *events.BufferedSubscription[events.Reward]
+	if request.Watch {
+		matcher := rewardsMatcher{request, ctx}
+		var err error
+		sub, err = events.SubscribeMatched(matcher.match)
+		if err != nil {
+			return status.Error(codes.Internal, err.Error())
+		}
+		defer sub.Close()
+		if err := stream.SendHeader(metadata.MD{}); err != nil {
+			return status.Errorf(codes.Unavailable, "can't send header")
+		}
+	}
+
+	dbChan := make(chan *types.Reward, 100)
+	errChan := make(chan error, 1)
+
+	ops, err := toRewardOperations(toRewardRequest(request))
+	if err != nil {
+		return status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	// send db data to chan to avoid buffer overflow
+	go func() {
+		defer close(dbChan)
+		if err := rewards.IterateRewardsOps(s.db, ops, func(rwd *types.Reward) bool {
+			select {
+			case dbChan <- rwd:
+				return true
+			case <-ctx.Done():
+				// exit if the stream context is cancelled
+				return false
+			}
+		}); err != nil {
+			select {
+			case errChan <- status.Error(codes.Internal, err.Error()):
+			default:
+				ctxzap.Error(stream.Context(), "unable to send error", zap.Error(err))
+				select {
+				case <-ctx.Done(): // check ctx.Done() to avoid blocking
+				default:
+				}
+			}
+			return
+		}
+	}()
+
+	var eventsOut <-chan events.Reward
+	var eventsFull <-chan struct{}
+	if sub != nil {
+		eventsOut = sub.Out()
+		eventsFull = sub.Full()
+	}
+
+	for {
+		select {
+		case rst := <-eventsOut:
+			err := stream.Send(&spacemeshv2alpha1.Reward{
+				Versioned: &spacemeshv2alpha1.Reward_V1{V1: toReward(&types.Reward{
+					Layer:       rst.Layer,
+					TotalReward: rst.Total,
+					LayerReward: rst.LayerReward,
+					Coinbase:    rst.Coinbase,
+					SmesherID:   rst.SmesherID,
+				})},
+			})
+			switch {
+			case errors.Is(err, io.EOF):
+				return nil
+			case err != nil:
+				return status.Error(codes.Internal, err.Error())
+			}
+		default:
+			select {
+			case rst := <-eventsOut:
+				err := stream.Send(&spacemeshv2alpha1.Reward{
+					Versioned: &spacemeshv2alpha1.Reward_V1{V1: toReward(&types.Reward{
+						Layer:       rst.Layer,
+						TotalReward: rst.Total,
+						LayerReward: rst.LayerReward,
+						Coinbase:    rst.Coinbase,
+						SmesherID:   rst.SmesherID,
+					})},
```

Comment:
I wonder if it would make sense to use `types.Reward` for events instead of `events.Reward` to avoid these conversions. These structrs are the same.